src/Lib/Element.php

```diff
@@ -14,185 +14,150 @@
 
     protected $elements = [];
 
+    protected $externalOptions = [];
+
+    protected $externalElements = [];
+
     /**
-     * Add an option to this element
-     * Options are generally string key/ mixed value pairs
-     *
-     * @param string $key
-     * @param mixed $value
+     * Return the class name of the element converted into snake_case
      *
-     * @return $this
+     * @return string
      */
-    final protected function addOption($key, $value)
+    public function getElementName()
     {
-        $this->options[$key] = $value;
-        return $this;
+        $classNameWithNamespace = get_class($this);
+        $classNameWithoutNamespace = substr($classNameWithNamespace, strrpos($classNameWithNamespace, '\\')+1);
+        preg_match_all('/[A-Z][a-z]+/', $classNameWithoutNamespace, $classNameParts);
```

Comment:
We could do. I'm pretty relaxed about validation here though, because  the values aren't dynamic or user input. We will be naming the classes in order to take advantage of this method, or overloading it if this is not feasible. A poorly named class is just as likely to break namespacing before we even get here.